### ISSUE TITLE
%?RESOURCES is empty from programs

### ISSUE THREAD

**Issue by dakkar** [rakudo/rakudo#4173]
## The Problem

The [documentation](https://docs.raku.org/language/variables#%?RESOURCES) says that "`%?RESOURCES` is a compile-time variable available to the code of a Distribution"

A program inside a distribution seems, to me, to be part of the distribution. I can run it with `CompUnit::RepositoryRegistry.run-script`, similarly to how I can load modules by name.

But it appears that `%?RESOURCES` is empty when used in a program, even an installed one (I understand that a random script on its own is not "part of a distribution", so I wouldn't expect it to work there).

If the current behaviour is the correct one, then the documentation may need to be clarified.

## Steps to reproduce

Given the minimal distribution at https://github.com/dakkar/raku-test/commit/860e0010829d359a600ffc6c4ae6ca5ea2e5d999

`zef install .` then run `raku-test`

## Expected Behavior

Something like:

```
/home/dakkar/.raku/resources/EB9ABB9B73B90C2CC34880E6707A23C58ECD2E9F.
/home/dakkar/.raku/resources/EB9ABB9B73B90C2CC34880E6707A23C58ECD2E9F.
```

## Actual Behavior

```
/home/dakkar/.raku/resources/EB9ABB9B73B90C2CC34880E6707A23C58ECD2E9F.
Nil
```

## Environment
* Operating system: linux
* Compiler version (`perl6 -v` or `raku -v`):
    Welcome to Rakudo(tm) v2020.12.
    Implementing the Raku(tm) programming language v6.d.
    Built on MoarVM version 2020.12.

**Comment by tbrowder** [comment]
What you have to do is define a sub INSIDE the module to use %?RESOURCES, export the sub, then call the exported sub from your script that 'uses yourmodule'.

**Comment by dakkar** [comment]
Yes, that's what my test does, that's why it prints the right path once.

The documentation, to me, implies that it's not necessary.

Either the documentation needs to be clearer, or there's a bug (a pretty low-importance one, yes)

**Comment by tbrowder** [comment]
This is a docs problem. Please file this issue there, preferably with a PR for the docs.

**Comment by ugexe** [comment]
You could do the bin-script-as-a-module trick, and just `use MyCLI;` so your bin-script-as-a-module has access to `%?RESOURCES`

For this to work scripts would probably have to be precompiled (they aren't). And even once they are I'm not sure how a script that *isnt* installed would do this -- if I do `raku bin/my_script` is it supposed to somehow figure out if that belongs to a local distribution and use that, or use a potentially installed version of the modules that script might use? What if it uses both? This is basically the same reason you don't get `%?RESOURCES` if you do `require "/path/to/some/module.rakumod"` instead of pointing at the root of the distribution.

Note I don't mean to just wave this issue away, I just have no idea how this could be pulled off. For instance I think a bigger issue is this also applies to `$?DISTRIBUTION`, such that a bin script cannot do e.g. `use MyModule:ver($?DISTRIBUTION.meta<version>)` to ensure that bin script loads only the module matching the version it was packaged with.

**Comment by niner** [comment]
I disagree. While the docs could indeed mention this situation, it's not exactly as planned. There's no reason things have to stay this way. Having support for `%?RESOURCES` in scripts was just never important enough for someone to actually implement it. But it's entirely possible and reasonable to want it. It's just that it would be quite a bit of work, maybe even tricky in some places and the workaround is as you described quite trivial.

**Comment by dakkar** [comment]
> You could do the bin-script-as-a-module trick, and just `use MyCLI;` so your bin-script-as-a-module has access to `%?RESOURCES`

yes, again, I did that in my test.

> And even once they are I'm not sure how a script that _isnt_ installed would do this -- if I do `raku bin/my_script` is it supposed to somehow figure out if that belongs to a local distribution and use that, or use a potentially installed version of the modules that script might use?

Yes, I wrote the same question in my first message, thank you for writing it more clearly.

Does `%?RESOURCES` work in non-installed *modules*? how do they know what distribution they belong to?

**Comment by dakkar** [comment]
Yes, `%?RESOURCES` works in non-installed modules:
```
~/src/raku-test$ raku -Ilib ./bin/raku-test
/home/dakkar/src/raku-test/resources/text
Nil
```

I guess that `-I` sets up a `CompUnit::Repository` or something like that, and `%?RESOURCES` is built from it.

**Comment by ugexe** [comment]
> yes, again, I did that in my test.

That is a different method tbrowder suggested; I see the 'add a method for accessing `%?RESOURCES` to one of the modules' in the test, but not actually putting the bin script code that would actually access `%?RESOURCES` inside a module ala https://github.com/ugexe/zef/blob/master/bin/zef
